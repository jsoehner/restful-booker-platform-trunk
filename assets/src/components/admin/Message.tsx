import React, { useEffect, useState } from 'react';
import ReactModal from 'react-modal';

interface MessageProps {
  messageId: number;
  closeMessage: () => void;
  refreshMessageList: () => void;
}

interface MessageDetails {
  id: number;
  name: string;
  email: string;
  phone: string;
  subject: string;
  description: string;
  read: boolean;
}

const Message: React.FC<MessageProps> = ({ messageId, closeMessage, refreshMessageList }) => {
  const [message, setMessage] = useState<MessageDetails | null>(null);

  useEffect(() => {
    const fetchMessage = async () => {
      try {
        const response = await fetch(`/api/message/${messageId}`);
        if (response.ok) {
          const data = await response.json();
          setMessage(data);
          
          // Mark as read if not already
          if (data && !data.read) {
            markAsRead();
          }
        }
      } catch (error) {
        console.error('Error fetching message:', error);
      }
    };

    fetchMessage();
  }, [messageId]);

  const markAsRead = async () => {
    try {
      const response = await fetch(`/api/message/${messageId}/read`, {
        method: 'PUT',
      });
      
      if (response.ok) {
        // Update the message list and count
        refreshMessageList();
        if (typeof window !== 'undefined' && (window as any).updateMessageCount) {
          (window as any).updateMessageCount();
        }
      }
    } catch (error) {
      console.error('Error marking message as read:', error);
    }
  };

  if (!message) {
    return <div></div>;
  }

  return (
    <ReactModal 
        isOpen={true}
        ariaHideApp={false}
        contentLabel="onRequestClose Example"
        className="message-modal" >
        
        <div data-testid="message">
            <div className="form-row">
                <div className="col-10">
                    <p><span>From: </span>{message.name}</p>
                </div>
                <div className="col-2">
                    <p><span>Phone: </span>{message.phone}</p>
                </div>
            </div>
            <div className="form-row">
                <div className="col-12">
                    <p><span>Email: </span>{message.email}</p>
                </div>
            </div>
            <div className="form-row">
                <div className="col-12">
                    <p><span>{message.subject}</span></p>
                </div>
            </div>
            <div className="form-row">
                <div className="col-12">
                    <p>{message.description}</p>
                </div>
            </div>
            <div className="form-row">
                <div className="col-12">
                    <button className="btn btn-outline-primary" onClick={() => closeMessage()}>Close</button>
                </div>
            </div>
        </div>
    </ReactModal>
  );
};

export default Message; 