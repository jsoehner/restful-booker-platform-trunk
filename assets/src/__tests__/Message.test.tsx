import React from 'react';
import Message from '../components/admin/Message';
import { render, waitFor, screen, act } from '@testing-library/react';

const messageData = {
  name: "Mark Winteringham",
  email: "mark@mwtestconsultancy.co.uk",
  phone: "01821 912812",
  subject: "Subject description here",
  description: "Lorem ipsum dolores est"
};

describe('Message Component', () => {
  beforeEach(() => {
    // Mock the GET request
    global.fetch = jest.fn()
      .mockImplementationOnce(() => Promise.resolve({
        ok: true,
        json: () => Promise.resolve(messageData)
      }))
      // Mock the PUT request for marking as read
      .mockImplementationOnce(() => Promise.resolve({
        ok: true
      }));
  });

  test('Message popup is populated with details', async () => {
    await act(async () => {
      render(<Message messageId={1} refreshMessageList={() => {}} closeMessage={() => {}} />);
    });

    await waitFor(() => {
      const modalComponent = screen.getByTestId(/message/);
      expect(modalComponent).toMatchSnapshot();
    });
  });
});
