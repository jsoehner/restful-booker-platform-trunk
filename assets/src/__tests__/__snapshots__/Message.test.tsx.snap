// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Component Message popup is populated with details 1`] = `
<div
  data-testid="message"
>
  <div
    class="form-row"
  >
    <div
      class="col-10"
    >
      <p>
        <span>
          From: 
        </span>
        Mark Winteringham
      </p>
    </div>
    <div
      class="col-2"
    >
      <p>
        <span>
          Phone: 
        </span>
        01821 912812
      </p>
    </div>
  </div>
  <div
    class="form-row"
  >
    <div
      class="col-12"
    >
      <p>
        <span>
          Email: 
        </span>
        mark@mwtestconsultancy.co.uk
      </p>
    </div>
  </div>
  <div
    class="form-row"
  >
    <div
      class="col-12"
    >
      <p>
        <span>
          Subject description here
        </span>
      </p>
    </div>
  </div>
  <div
    class="form-row"
  >
    <div
      class="col-12"
    >
      <p>
        Lorem ipsum dolores est
      </p>
    </div>
  </div>
  <div
    class="form-row"
  >
    <div
      class="col-12"
    >
      <button
        class="btn btn-outline-primary"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
